nvflare/app_opt/tf/fedopt_ctl.py

```diff
@@ -118,39 +118,53 @@ def update_model(self, global_model: FLModel, aggr_result: FLModel):
         specific set of Variables.
 
         """
-        # Get the Keras model stored in memory in persistor.
         global_model_tf = self.persistor.model
         global_params = global_model_tf.trainable_weights
+        num_trainable_weights = len(global_params)
 
-        # Compute model diff: need to use model diffs as
-        # gradients to be applied by the optimizer.
-        model_diff_params = {k: aggr_result.params[k] - global_model.params[k] for k in global_model.params}
-        model_diff = self._to_tf_params_list(model_diff_params, negate=True)
+        model_diff_params = {}
+
+        w_idx = 0
+
+        for key, param in global_model.params.items():
+            if w_idx >= num_trainable_weights:
+                break
 
-        # Apply model diffs as gradients, using the optimizer.
+            if param.shape == global_params[w_idx].shape:
+                model_diff_params[key] = (
+                    aggr_result.params[key] - param
+                    if aggr_result.params_type == ParamsType.FULL
+                    else aggr_result.params[key]
+                    )
+                w_idx += 1
+
+        model_diff = self._to_tf_params_list(model_diff_params, negate=True)
         start = time.time()
+
         self.optimizer.apply_gradients(zip(model_diff, global_params))
         secs = time.time() - start
 
-        # Convert updated global model weights to
-        # numpy format for FLModel.
         start = time.time()
         weights = global_model_tf.get_weights()
-        w_idx = 0
+
         new_weights = {}
-        for key in global_model.params:
-            w = weights[w_idx]
-            while global_model.params[key].shape != w.shape:
-                w_idx += 1
-                w = weights[w_idx]
-            new_weights[key] = w
-        secs_detach = time.time() - start
+        for w_idx, key in enumerate(global_model.params):
+            if key in model_diff_params:
+                new_weights[key] = weights[w_idx]
+
+            else:
 
+                new_weights[key] = (
+                    aggr_result.params[key]
+                    if aggr_result.params_type == ParamsType.FULL
+                    else global_model.params[key] + aggr_result.params[key]
+                )
```

Comment:
do you need the 
"if aggr_result.params[key].shape == global_model.params[key].shape:" ?

Reply:
Yes, since in the keys in aggr_result differ from the model layers, I need to identify the trainable variables within aggr_result.params[key]

Reply:
if you need that, could you add that condition to your PR? thanks

Reply:
Since I added this condition on line 133 of my script to separate the trainable variables and store them in the model_diff_param dictionary, I don't need to apply that logic again here. At this point, I am certain that the model_diff_param dictionary contains only the trainable variables.

Reply:
line 160, "global_model.params[key] + aggr_result.params[key]"

if the shape mismatch this statement will fail.

but if you are sure that global_model.params[key] will always have the same shape as  aggr_result.params[key]
then we are good.